### PR TITLE
[Fix #7064] Multiline when then cop

### LINKED ISSUE #7064: [Case statement] Disallow "then" keyword for multi-line "when" clauses
In a case statement we can use the `then` keyword to have the whole `when` clause in a single line, eg.

```ruby
case temperature
when (-30..0) then puts "Freezing!"
when (1..10) then puts "Cold"
when (11..20) then puts "Not bad"
when (21..28) then puts "Warm"
when (29..40) then puts "Quite hot"
else
  puts "Not on my scale"
end 
```

We can rewrite this case statement by putting all the `puts` calls in separate lines:

```ruby
case temperature
when (-30..0) then
  puts "Freezing!"
when (1..10) then 
  puts "Cold"
when (11..20) then 
  puts "Not bad"
when (21..28) then 
  puts "Warm"
when (29..40) then 
  puts "Quite hot"
else
  puts "Not on my scale"
end 
```

In this case, the `then` keyword is optional. I don't see any use cases when you would want to leave it there as it just generates more noise. Rubocop should complain about this and the correct version should be:

```ruby
case temperature
when (-30..0)
  puts "Freezing!"
when (1..10) 
  puts "Cold"
when (11..20) 
  puts "Not bad"
when (21..28) 
  puts "Warm"
when (29..40) 
  puts "Quite hot"
else
  puts "Not on my scale"
end 
```

### PR DESCRIPTION
**Pull request by okuramasafumi**
This cop check the use of the `then` keyword in multi-line when
statements.
Its name comes from `Style/MultilineIfThen` cop.
This cop is enabled by default because `Style/MultilineIfThen` cop is
enabled by default.
Related issue is #7064 

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 * [#5976](https://github.com/rubocop-hq/rubocop/issues/5976): Remove Rails cops. ([@koic][])
 * [#5976](https://github.com/rubocop-hq/rubocop/issues/5976): Remove `rubocop -R/--rails` option. ([@koic][])
+* [#7114](https://github.com/rubocop-hq/rubocop/pull/7114): Add `MultilineWhenThen` cop. ([@okuramasafumi][])
```
Comment (koic) [inline CHANGELOG.md]: This entry should be noted in `### New features`. And please squash to one commit this PR.

**Comment by koic** [comment]
`Style/MultilineIfThen` cop is based on the Ruby Style Guide.
https://github.com/rubocop-hq/ruby-style-guide#no-then

Only `if / unless` is mentioned above. `case` should be added to it. I think this change is appropriate if accepted there.

File: config/default.yml [hunk 2]
```diff
@@ -3020,6 +3020,11 @@ Style/MultilineTernaryOperator:
   Enabled: true
   VersionAdded: '0.9'
 
+Style/MultilineWhenThen:
+  Description: 'Do not use then for multi-line when statement'
```
Comment (koic) [inline config/default.yml]: Add `StyleGuide: '#no-then'` if The Ruby Style Guide accepts it.
https://github.com/rubocop-hq/rubocop/blob/v0.71.0/config/default.yml#L3458

File: lib/rubocop/cop/style/multiline_when_then.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      #   case str
+      #   when 'rubocop' then puts 'rubocop'
+      #   else
+      #     puts 'not rubocop'
+      #   end
+      #
+      class MultilineWhenThen < Cop
+        include RangeHelp
+
+        MSG = 'Do not use then for multiline when statement.'
```
Comment (bbatsov) [inline lib/rubocop/cop/style/multiline_when_then.rb]: Please, surround `then` and `when` in backticks.

Comment (okuramasafumi) [reply lib/rubocop/cop/style/multiline_when_then.rb]: OK, I'll fix it soon.

File: config/default.yml [hunk 4]
```diff
@@ -3020,6 +3020,12 @@ Style/MultilineTernaryOperator:
...
```
Comment (koic) [inline config/default.yml]: Please add a period at EOL.

```diff
- Description: 'Do not use then for multi-line when statement'
+ Description: 'Do not use then for multi-line when statement.'
```

File: lib/rubocop/cop/style/multiline_when_then.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      #     puts 'not rubocop'
+      #   end
+      #
+      #   # good
+      #   case str
+      #   when 'rubocop' then puts 'rubocop'
+      #   else
+      #     puts 'not rubocop'
+      #   end
+      #
```
Comment (koic) [inline lib/rubocop/cop/style/multiline_when_then.rb]: How about the following example code based on https://github.com/rubocop-hq/ruby-style-guide/pull/752?

```ruby
# bad
case foo
when bar then
end

# good
case foo
when bar
end

# good
case foo
when bar then do_something
end
```

Comment (okuramasafumi) [reply lib/rubocop/cop/style/multiline_when_then.rb]: Right.

File: lib/rubocop/cop/style/multiline_when_then.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+      #   # good
+      #   case foo
+      #   when bar
+      #   end
+
+      #   # good
+      #   case foo
+      #   when bar then do_something
+      #   end
```
Comment (koic) [inline lib/rubocop/cop/style/multiline_when_then.rb]: It needs to be connected with `#`.

```diff
      # @example
      #   # bad
      #   case foo
      #   when bar then
      #   end
-
+     #
      #   # good
      #   case foo
      #   when bar
      #   end
-
+     #
      #   # good
      #   case foo
      #   when bar then do_something
      #   end
```

Some of example code generated for document is lost.
https://github.com/rubocop-hq/rubocop/pull/7114/files#diff-3f2ebe9c9435b29358487264049625a4

File: spec/rubocop/cop/style/multiline_when_then_spec.rb [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    RUBY
+    expect(new_source).to eq(<<~RUBY)
+      case str
+      when 'rubocop'
+        puts 'rubocop'
+      else
+        'not rubocop'
+      end
+    RUBY
+  end
```
Comment (koic) [inline spec/rubocop/cop/style/multiline_when_then_spec.rb]: Can you update test code using `foo` and `bar` according to example code?

**Comment by bbatsov** [comment]
@okuramasafumi The specs are currently failing. You'll also have to rebase on top of `master`.

**Comment by okuramasafumi** [comment]
@bbatsov Thanks, I'll fix and rebase it when I have time (hopefully next week).

**Comment by okuramasafumi** [comment]
@bbatsov I rebased on master and fixed the failing specs. I also added some specs so that the cop is more robust.

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**